Redesign CLI Plugin

i-am-offline opened the pull request:
I did some redesign on the CLI Plugin:

- Change into SmartPlugin
- Require passwort to logon
- Prepare for other plugins to add own commands (similar to my sh.py pull request 181)

For the password I added a function "create_hash" to the general SmartHomeNG Tools class. It should be possible to use this function for the hashed password in the backend plugin, too (#101)

i-am-offline commented:
I found two old push requests for sh.py which were not included in SmartHomeNG (mknx/smarthome#148, mknx/smarthome#148). They are included, now.

ohinckel commented:
Can we avoid merging different PR **manually** into a new PR? According to #52 @cstrassburg already merged the stuff for the new CLI commands (sl, si, st, ld) and we could also re-use the PR for the other feature (add commands by other plugins) by simply creating a new PR in this project.

In the end it makes it complicated when some have already merged the features into his private branch. They are different commits and will result in a conflict when HEAD will be merged into this private branch.

I also like it having one PR for one feature. But it is only MHO :wink:

msinn commented:
My first thoughts:
I think tools.py is on the way out (at least that's what cstrassburg planned). The hash function should go to utils.py and all functions in utils.py should have unit tests.

The modifications are numerous and quite complex. I am gonna have a deeper look into them.

i-am-offline commented:
I saw #52 but there is no branch or something else where I could base my PR on, so I used the current develop branch to have it mergeable. If there's something else let me know and I will try to rebase these commits.

Regarding tools.py/utils.py: It sould be quiet easy to move the new function from tools.py to utils.py and create the required unit tests. I can do so when the it's clear where to base the PR on.

ohinckel commented:
> there is no branch or something else where I could base my PR on

You can access all public repos on github (at least read-only). So you can just fetch the changes from my branch and let your work base on this.

I would update and merge my branch, but @cstrassburg already merged something somewhere. Unfortunaltely he is not responsive on my questions and I don't know the current status of this PR. Somebody else managed to get in touch with @cstrassburg ?

i-am-offline commented:
Let me just check I will perform the right steps: 

- I would add your repository as another remote to my local repository
- Reset my feature-cli-plugin to the latest develop branch
- merge your two (?) dev-cli* branches
- cherry-pick my comits, probably resolving conflicts
- push everything again into my github repository
I think this would require a forced push. So it might be better to create a different branch, but I don't know if I can change the PR to use the new branch instead? Of course I could close this pr and create a new one

ohinckel commented:
@i-am-offline, yes this sound really great, thanks! And I think it is not a problem doing a forced push since it is your branch and nobody other will use it currently (guess).

Maybe one more steps: move the stuff from `tools.py` into `utils.py`. Create tests for this.

ohinckel commented:
I just updated my two branches with the current develop version. Branches can be found [here](https://github.com/ohinckel/smarthome/tree/dev-cli-scheduler) and (there)[https://github.com/ohinckel/smarthome/tree/dev-cli-log-dump]. Both can be merged without conflict now.

i-am-offline commented:
OK, I'm done for now.

ohinckel commented on plugins/cli/README.md:
```diff
@@ -12,20 +12,22 @@ plugin.conf
 <code>port =</code> used network port, default 2323
 <code>update =</code> restrict the access of the items to read only (false, default) or allows read/write access (true)
-
+<code>hashed_password = </code> pasword that needs to be entered on login. SHA-512 hashed. Value shown above is "very_secure_password"
```
Typo in "password". Do we want to use sha512password instead of hashed_password? In #101 we discussed the hashed password support a little bit and there it is named md5password (so one see which hash algorithm is used). I think we should be consistent in both plugins.

i-am-offline replied:
I choose "hashed_password" to be independent from the used hash function. In the latest commit in #101 i changed the parameter name there also to "hashed_password". As with #101 it is possible to create the hashes using the backend plugin, there I would not bother the users with the type of the hash function. I would also change the backend plugin so that it uses the new hash method in Utils.py.

ohinckel commented on plugins/cli/__init__.py:
```diff
@@ -22,234 +23,548 @@
+
+        if self.__wait_for_password:
+            # We are waiting for a password. So we take the entered command as password
+            if Utils.create_hash(cmd) == self.hashed_password.lower():
```
Comparing a password could also be a method in utils.py.

i-am-offline replied:
You are right. It should be a method in utils.py. Not because the ".lower()" part might be forgotten by someone else, but because other checks could be made in there as well. E.g.:

- Empty password -> return "not matching"
- Password shorter than [insert min length here] characters -> return "not matching"
- ...

ohinckel commented on plugins/cli/__init__.py:
```diff
@@ -22,234 +23,548 @@
+        self.logger = logging.getLogger(__name__)
+
+        if hashed_password is None or hashed_password == '':
+            self.logger.error("CLI: You must set a password to use this plugin. CLI Plugin will not be initialized!")
```
Do we really want to make the password setting mandatory? All instalation will fail to start the plugin when updating sh.py to the new CLI plugin version.

I prefer to
a) log a message that one of the future release will require a password configured **or** 
b) make it optional and add a recommendation to the plugin readme to setup a password.

i-am-offline replied:
I knew that this one would cause a discussion. ;-))
In my opinion, security should be mandatory, this is why I made it this way. I think it's hard but it's reasonable. However I see your point that after an update the CLI plugin will not start and the reason for this might be somewhere deep in the log files.
How about option 
c) CLI Plugin will (in addition to the error message) start, but not allow to login, instead show a message and close the connection.

ohinckel commented on plugins/cli/__init__.py:
```diff
@@ -22,234 +23,548 @@
+
+        lib.connection.Server.__init__(self, ip, port)
+        self.sh = smarthome
+        self.updates_allowed = smarthome.string2bool(update)
```
Can we use `Utils.to_bool()` instead of `smarthome.string2bool()`?

i-am-offline replied:
That's one of the lines I did not touch ... at least not yet

ohinckel commented:
Thanks for your really great refactoring! I added some comments to the code...

msinn commented:
About a mandatory password: I am against it. I don't need it in my environment. 
I like to be able to define a password but I don't like to be forced to have one.

psilo909 commented:
same for me. we should also think about the (optional) possibility to configure a password at ONE place for smarthomeng and use it throughout different plugins! later on, we can change that into a user and role system..

i-am-offline commented:
Seems I've been overruled with the mandatory password.
I would change it in the following way:

- Password is optional
- If hashed_password is not set, a warning will be written into the log
- If hashed_password is set to "none" instead of a hash value, even the warning will be supressed
- Default will be that hashed_password is not set

Having a general admin user/password for SmartHomeNG is also not a bad idea. However, I think that this requires changes at several plugins, so it should be separated from this pr. What I can do here with regards to such a change is, that I'm implementing an username, too

ohinckel commented:
Your proposed changes are OK for me. And I would also not start introducing **one** password for more stuff and prefer the role based system - but this would be another issue, since it is just to complex to implement this now within this PR.

When thining about the role based system it would be great when we have something we could use in other plugins too (e.g. the visu_acl or nw_acl or ....). But we could start thinking about this feature and collect stuff in a new issue.

msinn commented:
@i-am-offline I like your ideas about the implementation of the "optional password".

About the role-based model, I am with ohinckel. We should collect requirements and ideas and create an authorization system which can be used by all plugins.

i-am-offline commented:
So, I did the changes regarding the password and some other changes as well. 

- I added a check_hashed_password method to utils.py as @ohinckel suggested. I also added a method "is_hash" to check if something is a valid hash.
- I used those methods for the CLI Plugin
- I did some code cleanup
- Finaly I added the processing of the client reply on the ECHO signals that was previously missing.

From my point of view the redesign of the CLI plugin is now complete.

psilo909 commented:
should be merged by someone, who uses the plugin :)

ohinckel commented:
I'll merge and test this when I have more time in the next days.

ohinckel commented on plugins/cli/__init__.py:
```diff
@@ -22,234 +23,621 @@
+        else:
+            logs = self.sh.return_logs()
+            if parameter not in logs:
+                self.push("Log '{0}' does not exist\n".format(parameter))
```
This should be ´handler.push()´ instead of ´self.push()´.

i-am-offline commented:
I've fixed this. Thank you for testing.

ohinckel commented on plugins/cli/__init__.py:
```diff
@@ -22,234 +23,621 @@
+            for entry in log.last(10):
+                values = [str(value) for value in entry]
+                handler.push(str(values))
+        handler.push("\n")
```
The newline character should be on the end of each log line and not on the end of the whole output. So changing the indenting to 16 character would fix this output "error". :wink:
